### DISCUSSION TITLE
Incorrect generation of openapi.json with duplicated name Enum class

### DISCUSSION THREAD

**Discussion by mnixry** [fastapi/fastapi#6954] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

#### `main.py`

```Python
from fastapi import FastAPI

import router1
import router2

app = FastAPI()

app.include_router(router1.router, prefix="/r1")
app.include_router(router2.router, prefix="/r2")
```

#### `router1.py`

```python
from enum import Enum

from fastapi import APIRouter


class DuplicatedEnum(str, Enum):
    value1 = "value1"


router = APIRouter()


@router.get("/")
async def _(type: DuplicatedEnum = DuplicatedEnum.value1):
    pass
```

#### `router2.py`

```python
from enum import Enum

from fastapi import APIRouter


class DuplicatedEnum(str, Enum):
    value2 = "value2"


router = APIRouter()


@router.get("/")
async def _(type: DuplicatedEnum = DuplicatedEnum.value2):
    pass
```

### Description

* Open the browser and call the document `/redoc`.
* It raised error `Error resolving $ref pointer "http://127.0.0.1:8000/openapi.json#/components/schemas/DuplicatedEnum". Token "DuplicatedEnum" does not exist.`
* After I checked `openapi.json`, it gives wrong parameter schema:

```jsonc
{
	"openapi": "3.0.2",
	"info": {
		"title": "FastAPI",
		"version": "0.1.0"
	},
	"paths": {
		"/r1/": {
			"get": {
				"summary": " ",
				"operationId": "__r1__get",
				"parameters": [{
					"required": false,
					"schema": {
						"allOf": [{
							"$ref": "#/components/schemas/DuplicatedEnum" //should be `router1__DuplicatedEnum`
						}],
						"default": "value1"
					},
					"name": "type",
					"in": "query"
				}],
				"responses": {
					"200": {
						"description": "Successful Response",
						"content": {
							"application/json": {
								"schema": {}
							}
						}
					},
					"422": {
						"description": "Validation Error",
						"content": {
							"application/json": {
								"schema": {
									"$ref": "#/components/schemas/HTTPValidationError"
								}
							}
						}
					}
				}
			}
		},
		"/r2/": {
			"get": {
				"summary": " ",
				"operationId": "__r2__get",
				"parameters": [{
					"required": false,
					"schema": {
						"allOf": [{
							"$ref": "#/components/schemas/DuplicatedEnum"
						}],
						"default": "value2"
					},
					"name": "type",
					"in": "query"
				}],
				"responses": {
					"200": {
						"description": "Successful Response",
						"content": {
							"application/json": {
								"schema": {}
							}
						}
					},
					"422": {
						"description": "Validation Error",
						"content": {
							"application/json": {
								"schema": {
									"$ref": "#/components/schemas/HTTPValidationError"
								}
							}
						}
					}
				}
			}
		}
	},
	"components": {
		"schemas": {
			"HTTPValidationError": {
				"title": "HTTPValidationError",
				"type": "object",
				"properties": {
					"detail": {
						"title": "Detail",
						"type": "array",
						"items": {
							"$ref": "#/components/schemas/ValidationError"
						}
					}
				}
			},
			"ValidationError": {
				"title": "ValidationError",
				"required": ["loc", "msg", "type"],
				"type": "object",
				"properties": {
					"loc": {
						"title": "Location",
						"type": "array",
						"items": {
							"type": "string"
						}
					},
					"msg": {
						"title": "Message",
						"type": "string"
					},
					"type": {
						"title": "Error Type",
						"type": "string"
					}
				}
			},
			"router1__DuplicatedEnum": {
				"title": "DuplicatedEnum",
				"enum": ["value1"],
				"type": "string",
				"description": "An enumeration."
			},
			"router2__DuplicatedEnum": {
				"title": "DuplicatedEnum",
				"enum": ["value2"],
				"type": "string",
				"description": "An enumeration."
			}
		}
	}
}
```

### Environment

* OS: [e.g. Linux / Windows / macOS]: Linux
* FastAPI Version [e.g. 0.3.0]: 0.63.0
* Python version: 3.9.1

**Comment by falkben** [comment]
Can confirm. Workaround would be to name the the Enum classes differently. Can you submit a PR to fix?

**Comment by mnixry** [comment]
Confirmed, this is the `pydantic` schema generate issue ( https://github.com/samuelcolvin/pydantic/issues/1857 )
However it has been fixed in  https://github.com/samuelcolvin/pydantic/pull/2226 https://github.com/samuelcolvin/pydantic/commit/13a5c7d676167b415080de5e6e6a74bea095b239 which have not been released yet. :cry: 

As a temporary solution, I have to degrade pydantic to [1.5.1](https://github.com/samuelcolvin/pydantic/releases/tag/v1.5.1) to avoid this problem.

**Comment by mnixry** [comment, marked as answer]
[PyDantic 1.8 Released](https://github.com/samuelcolvin/pydantic/releases/tag/v1.8) which contains this fix

**Comment by tiangolo** [comment]
Thanks for the help here @falkben ! 👏  🙇 

Thanks for reporting back and closing the issue @mnixry  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
